core/src/main/java/com/linecorp/armeria/server/file/FileService.java

```diff
@@ -257,35 +258,18 @@ private HttpFile findFile(ServiceRequestContext ctx, HttpRequest req) {
                     });
                 });
             } else {
-                // Redirect to the slash appended path if:
-                // 1) /index.html exists or
-                // 2) it has a directory listing.
-                final String indexPath = decodedMappedPath + "/index.html";
-                return findFile(ctx, indexPath, encodings, decompress).thenCompose(indexFile -> {
-                    if (indexFile != null) {
-                        return UnmodifiableFuture.completedFuture(true);
-                    }
+                final List<String> fallbackExtensions = config.fallbackFileExtensions();
+                if (fallbackExtensions.isEmpty()) {
+                    return findFileWithIndexPath(ctx, decodedMappedPath, encodings, decompress);
+                }
 
-                    if (!config.autoIndex()) {
-                        return UnmodifiableFuture.completedFuture(false);
-                    }
-
-                    return config.vfs().canList(ctx.blockingTaskExecutor(), decodedMappedPath);
-                }).thenApply(canList -> {
-                    if (canList) {
-                        try (TemporaryThreadLocals ttl = TemporaryThreadLocals.acquire()) {
-                            final StringBuilder locationBuilder = ttl.stringBuilder()
-                                                                     .append(ctx.path())
-                                                                     .append('/');
-                            if (ctx.query() != null) {
-                                locationBuilder.append('?')
-                                               .append(ctx.query());
-                            }
-                            return HttpFile.ofRedirect(locationBuilder.toString());
-                        }
-                    } else {
-                        return HttpFile.nonExistent();
+                // Try appending file extensions if it was a file access and file extensions are configured.
+                return findFileWithExtensions(ctx, fallbackExtensions.iterator(), decodedMappedPath,
```

Comment:
Note: I understood the order of 1) check extensions 2) and then check `index.html` is important since the objective is to avoid a trailing slash